File: typescript/akm.ts
/**
 * AKM Binary Format — TypeScript Reference Implementation
 *
 * Encoder and decoder for the .akm (Adelkrieg Map) binary format.
 * Uses DataView for big-endian I/O and Web Crypto API for SHA-256.
 * Works in both Node.js and browsers with no external dependencies.
 */

// ============================================================================
// Types
// ============================================================================

export enum BlendMode {
	Normal = 0,
	Multiply = 1,
	Screen = 2,
	Overlay = 3,
}

export enum ImageFormat {
	PNG = 0,
	WebP = 1,
	AVIF = 2,
}

export interface AkmAuthor {
	readonly authorId: string; // UUID string
	readonly name: string;
}

export interface AkmRing {
	/** Alternating [x0, y0, x1, y1, ...] coordinates. Length must be even and >= 6. */
	readonly points: Float32Array;
}

export interface AkmRegion {
	/** Ring 0 = exterior (CCW), rings 1..N = holes (CW). */
	readonly rings: readonly AkmRing[];
}

export interface AkmTerritory {
	readonly territoryId: string; // UUID string
	readonly label: string;
	readonly centroidX: number;
	readonly centroidY: number;
	/** Multiple regions = exclaves. */
	readonly regions: readonly AkmRegion[];
}

export interface AkmEdge {
	readonly sourceId: string; // UUID string
	readonly targetId: string; // UUID string
	readonly bidirectional: boolean;
}

export interface AkmLayer {
	readonly name: string;
	readonly zOrder: number;
	readonly blendMode: BlendMode;
	/** 0-255, maps to 0.0-1.0 alpha. */
	readonly opacity: number;
	readonly imageFormat: ImageFormat;
	readonly imageData: Uint8Array;
}

export interface AkmMap {
	readonly mapId: string; // UUID string
	readonly canvasWidth: number;
	readonly canvasHeight: number;
	readonly displayName: string;
	readonly description: string;
	readonly authors: readonly AkmAuthor[];
	/** Unix timestamp in seconds. */
	readonly createdAt: number;
	readonly editorVersion: string;
	readonly territories: readonly AkmTerritory[];
	readonly edges: readonly AkmEdge[];
	readonly layers: readonly AkmLayer[];
}

// ============================================================================
// Constants
// ============================================================================

const SIGNATURE = new Uint8Array([0x83, 0x41, 0x4b, 0x4d, 0x0d, 0x0a, 0x1a, 0x0a]);
const HEADER_SIZE = 16;

const CHUNK_META = encodeAscii("META");
const CHUNK_GEOM = encodeAscii("GEOM");
const CHUNK_ADJC = encodeAscii("ADJC");
const CHUNK_LAYR = encodeAscii("LAYR");
const CHUNK_CKSM = encodeAscii("CKSM");

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder("utf-8");

// ============================================================================
// Encoder
// ============================================================================

export async function encode(map: AkmMap): Promise<Uint8Array> {
	const chunks: Uint8Array[] = [];

	// File header (16 bytes)
	const header = new Uint8Array(HEADER_SIZE);
	header.set(SIGNATURE);
	const hv = new DataView(header.buffer);
	hv.setUint16(8, 1); // version major
	hv.setUint16(10, 0); // version minor
	// bytes 12-15 reserved (already 0)
	chunks.push(header);

	// Chunks
	chunks.push(writeChunk(CHUNK_META, encodeMeta(map)));
	chunks.push(writeChunk(CHUNK_GEOM, encodeGeom(map)));
	chunks.push(writeChunk(CHUNK_ADJC, encodeAdjc(map)));
	for (const layer of map.layers) {
		chunks.push(writeChunk(CHUNK_LAYR, encodeLayer(layer)));
	}

	// Checksum
	const contentSoFar = concat(chunks);
	const hash = new Uint8Array(await crypto.subtle.digest("SHA-256", contentSoFar.buffer as ArrayBuffer));
	const cksmPayload = new Uint8Array(1 + hash.length);
	cksmPayload[0] = 0; // algorithm = SHA-256
	cksmPayload.set(hash, 1);
	chunks.push(writeChunk(CHUNK_CKSM, cksmPayload));

	return concat(chunks);
}

function encodeMeta(map: AkmMap): Uint8Array {
	const w = new BinaryWriter();
	w.uuid(map.mapId);
	w.u32(map.canvasWidth);
	w.u32(map.canvasHeight);
	w.string(map.displayName);
	w.string(map.description);
	w.u16(map.authors.length);
	for (const author of map.authors) {
		w.uuid(author.authorId);
		w.string(author.name);
	}
	w.i64(map.createdAt);
	w.string(map.editorVersion);
	return w.toBytes();
}

function encodeGeom(map: AkmMap): Uint8Array {
	const w = new BinaryWriter();
	w.u32(map.territories.length);
	for (const territory of map.territories) {
		w.uuid(territory.territoryId);
		w.string(territory.label);
		w.f32(territory.centroidX);
		w.f32(territory.centroidY);
		w.u16(territory.regions.length);
		for (const region of territory.regions) {
			w.u16(region.rings.length);
			for (const ring of region.rings) {
				const pointCount = ring.points.length / 2;
				w.u32(pointCount);
				for (let i = 0; i < ring.points.length; i++) {
					w.f32(ring.points[i]);
				}
			}
		}
	}
	return w.toBytes();
}

function encodeAdjc(map: AkmMap): Uint8Array {
	const w = new BinaryWriter();
	w.u32(map.edges.length);
	for (const edge of map.edges) {
		w.uuid(edge.sourceId);
		w.uuid(edge.targetId);
		w.u8(edge.bidirectional ? 0x01 : 0x00);
	}
	return w.toBytes();
}

function encodeLayer(layer: AkmLayer): Uint8Array {
	const w = new BinaryWriter();
	w.string(layer.name);
	w.u16(layer.zOrder);
	w.u8(layer.blendMode);
	w.u8(layer.opacity);
	w.u8(layer.imageFormat);
	w.u32(layer.imageData.length);
	w.bytes(layer.imageData);
	return w.toBytes();
}

function writeChunk(tag: Uint8Array, payload: Uint8Array): Uint8Array {
	const chunk = new Uint8Array(8 + payload.length);
	chunk.set(tag, 0);
	new DataView(chunk.buffer).setUint32(4, payload.length);
	chunk.set(payload, 8);
	return chunk;
}

// ============================================================================
// Decoder
// ============================================================================

export async function decode(data: Uint8Array): Promise<AkmMap> {
	if (data.length < HEADER_SIZE) {
		throw new Error("Data too short to be a valid .akm file");
	}

	const r = new BinaryReader(data);

	// File header
	const sig = r.bytes(8);
	if (!arraysEqual(sig, SIGNATURE)) {
		throw new Error("Invalid AKM file signature");
	}

	const versionMajor = r.u16();
	const versionMinor = r.u16();
	if (versionMajor > 1) {
		throw new Error(`Unsupported AKM version: ${versionMajor}.${versionMinor}`);
	}

	r.skip(4); // reserved

	// Chunks
	let mapId: string | null = null;
	let canvasWidth = 0;
	let canvasHeight = 0;
	let displayName: string | null = null;
	let description = "";
	let authors: AkmAuthor[] | null = null;
	let createdAt = 0;
	let editorVersion = "";
	let territories: AkmTerritory[] | null = null;
	let edges: AkmEdge[] | null = null;
	const layers: AkmLayer[] = [];

	let foundChecksum = false;

	while (r.remaining() > 0) {
		const tagBytes = r.bytes(4);
		const payloadSize = r.u32();
		if (payloadSize < 0) throw new Error("Negative chunk size");
		const payload = r.bytes(payloadSize);

		const tag = decodeAscii(tagBytes);

		if (tag === "CKSM") {
			if (payload.length < 1) throw new Error("CKSM chunk too short");
			if (payload[0] !== 0) throw new Error(`Unsupported checksum algorithm: ${payload[0]}`);
			const expectedHash = payload.slice(1);
			const contentBeforeCksm = data.slice(0, r.offset - 8 - payloadSize);
			const actualHash = new Uint8Array(await crypto.subtle.digest("SHA-256", contentBeforeCksm.buffer as ArrayBuffer));
			if (!arraysEqual(expectedHash, actualHash)) throw new Error("Checksum verification failed");
			foundChecksum = true;
			break;
		}

		const p = new BinaryReader(payload);

		switch (tag) {
			case "META": {
				mapId = p.uuid();
				canvasWidth = p.u32();
				canvasHeight = p.u32();
				displayName = p.string();
				description = p.string();
				const authorCount = p.u16();
				authors = [];
				for (let i = 0; i < authorCount; i++) {
					authors.push({ authorId: p.uuid(), name: p.string() });
				}
				createdAt = p.i64();
				editorVersion = p.string();
				break;
			}
			case "GEOM": {
				const territoryCount = p.u32();
				territories = [];
				for (let t = 0; t < territoryCount; t++) {
					territories.push(readTerritory(p));
				}
				break;
			}
			case "ADJC": {
				const edgeCount = p.u32();
				edges = [];
				for (let e = 0; e < edgeCount; e++) {
					const sourceId = p.uuid();
					const targetId = p.uuid();
					const bidirectional = (p.u8() & 0x01) !== 0;
					edges.push({ sourceId, targetId, bidirectional });
				}
				break;
			}
			case "LAYR": {
				const name = p.string();
				const zOrder = p.u16();
				const blendMode = p.u8() as BlendMode;
				const opacity = p.u8();
				const imageFormat = p.u8() as ImageFormat;
				const imageDataLength = p.u32();
				const imageData = p.bytes(imageDataLength);
				layers.push({ name, zOrder, blendMode, opacity, imageFormat, imageData });
				break;
			}
			// Unknown chunks are silently skipped (forward compatibility)
		}
	}

	if (!foundChecksum) throw new Error("Missing CKSM chunk");
	if (mapId == null || displayName == null || territories == null || edges == null || authors == null) {
		throw new Error("Missing required chunks (META, GEOM, ADJC)");
	}

	return {
		mapId,
		canvasWidth,
		canvasHeight,
		displayName,
		description,
		authors,
		createdAt,
		editorVersion,
		territories,
		edges,
		layers,
	};
}

function readTerritory(p: BinaryReader): AkmTerritory {
	const territoryId = p.uuid();
	const label = p.string();
	const centroidX = p.f32();
	const centroidY = p.f32();
	const regionCount = p.u16();
	const regions: AkmRegion[] = [];
	for (let r = 0; r < regionCount; r++) {
		const ringCount = p.u16();
		const rings: AkmRing[] = [];
		for (let k = 0; k < ringCount; k++) {
			const pointCount = p.u32();
			const points = new Float32Array(pointCount * 2);
			for (let i = 0; i < points.length; i++) {
				points[i] = p.f32();
			}
			rings.push({ points });
		}
		regions.push({ rings });
	}
	return { territoryId, label, centroidX, centroidY, regions };
}

// ============================================================================
// Binary Writer
// ============================================================================

class BinaryWriter {
	private chunks: Uint8Array[] = [];

	u8(value: number): void {
		this.chunks.push(new Uint8Array([value & 0xff]));
	}

	u16(value: number): void {
		const buf = new Uint8Array(2);
		new DataView(buf.buffer).setUint16(0, value);
		this.chunks.push(buf);
	}

	u32(value: number): void {
		const buf = new Uint8Array(4);
		new DataView(buf.buffer).setUint32(0, value);
		this.chunks.push(buf);
	}

	i64(value: number): void {
		const buf = new Uint8Array(8);
		const view = new DataView(buf.buffer);
		// Split into high and low 32-bit parts (big-endian)
		view.setInt32(0, Math.floor(value / 0x100000000));
		view.setUint32(4, value >>> 0);
		this.chunks.push(buf);
	}

	f32(value: number): void {
		const buf = new Uint8Array(4);
		new DataView(buf.buffer).setFloat32(0, value);
		this.chunks.push(buf);
	}

	uuid(value: string): void {
		const hex = value.replace(/-/g, "");
		if (hex.length !== 32) throw new Error(`Invalid UUID: ${value}`);
		const buf = new Uint8Array(16);
		for (let i = 0; i < 16; i++) {
			buf[i] = parseInt(hex.substring(i * 2, i * 2 + 2), 16);
		}
		this.chunks.push(buf);
	}

	string(value: string): void {
		if (!value || value.length === 0) {
			this.u16(0);
			return;
		}
		const encoded = textEncoder.encode(value);
		this.u16(encoded.length);
		this.chunks.push(encoded);
	}

	bytes(data: Uint8Array): void {
		this.chunks.push(data);
	}

	toBytes(): Uint8Array {
		return concat(this.chunks);
	}
}

// ============================================================================
// Binary Reader
// ============================================================================

class BinaryReader {
	private readonly data: Uint8Array;
	private readonly view: DataView;
	private pos: number = 0;

	constructor(data: Uint8Array) {
		this.data = data;
		this.view = new DataView(data.buffer, data.byteOffset, data.byteLength);
	}

	get offset(): number {
		return this.pos;
	}

	remaining(): number {
		return this.data.length - this.pos;
	}

	skip(n: number): void {
		this.ensure(n);
		this.pos += n;
	}

	u8(): number {
		this.ensure(1);
		return this.data[this.pos++];
	}

	u16(): number {
		this.ensure(2);
		const value = this.view.getUint16(this.pos);
		this.pos += 2;
		return value;
	}

	u32(): number {
		this.ensure(4);
		const value = this.view.getUint32(this.pos);
		this.pos += 4;
		return value;
	}

	i64(): number {
		this.ensure(8);
		const high = this.view.getInt32(this.pos);
		const low = this.view.getUint32(this.pos + 4);
		this.pos += 8;
		return high * 0x100000000 + low;
	}

	f32(): number {
		this.ensure(4);
		const value = this.view.getFloat32(this.pos);
		this.pos += 4;
		return value;
	}

	uuid(): string {
		const raw = this.bytes(16);
		const hex = Array.from(raw, (b) => b.toString(16).padStart(2, "0")).join("");
		return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
	}

	string(): string {
		const len = this.u16();
		if (len === 0) return "";
		return textDecoder.decode(this.bytes(len));
	}

	bytes(n: number): Uint8Array {
		this.ensure(n);
		const slice = this.data.slice(this.pos, this.pos + n);
		this.pos += n;
		return slice;
	}

	private ensure(n: number): void {
		if (this.pos + n > this.data.length) {
			throw new Error(`Unexpected end of data (need ${n} bytes at offset ${this.pos}, have ${this.data.length - this.pos})`);
		}
	}
}

// ============================================================================
// Utilities
// ============================================================================

function concat(arrays: Uint8Array[]): Uint8Array {
	let totalLength = 0;
	for (const arr of arrays) totalLength += arr.length;
	const result = new Uint8Array(totalLength);
	let offset = 0;
	for (const arr of arrays) {
		result.set(arr, offset);
		offset += arr.length;
	}
	return result;
}

function arraysEqual(a: Uint8Array, b: Uint8Array): boolean {
	if (a.length !== b.length) return false;
	for (let i = 0; i < a.length; i++) {
		if (a[i] !== b[i]) return false;
	}
	return true;
}

function encodeAscii(s: string): Uint8Array {
	const buf = new Uint8Array(s.length);
	for (let i = 0; i < s.length; i++) buf[i] = s.charCodeAt(i);
	return buf;
}

function decodeAscii(buf: Uint8Array): string {
	let s = "";
	for (let i = 0; i < buf.length; i++) s += String.fromCharCode(buf[i]);
	return s;
}

File: kotlin/src/akm/Akm.kt
/**
 * AKM Binary Format — Kotlin Reference Implementation
 *
 * Encoder and decoder for the .akm (Adelkrieg Map) binary format.
 * Zero external dependencies — uses only java.io, java.nio, java.security, java.util.
 */
package akm

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.time.Instant
import java.util.UUID

// ============================================================================
// Data Model
// ============================================================================

data class AkmMap(
    val mapId: UUID,
    val canvasWidth: Int,
    val canvasHeight: Int,
    val displayName: String,
    val description: String = "",
    val authors: List<AkmAuthor>,
    val createdAt: Instant?,
    val editorVersion: String = "",
    val territories: List<AkmTerritory>,
    val edges: List<AkmEdge> = emptyList(),
    val layers: List<AkmLayer> = emptyList(),
) {
    init {
        require(canvasWidth > 0 && canvasHeight > 0) { "Canvas dimensions must be positive" }
        require(displayName.isNotBlank()) { "Display name must not be blank" }
        require(authors.isNotEmpty()) { "At least one author is required" }
        require(territories.isNotEmpty()) { "At least one territory is required" }
    }
}

data class AkmAuthor(val authorId: UUID, val name: String) {
    init {
        require(name.isNotBlank()) { "Author name must not be blank" }
    }
}

data class AkmTerritory(
    val territoryId: UUID,
    val label: String = "",
    val centroidX: Float,
    val centroidY: Float,
    val regions: List<AkmRegion>,
) {
    init {
        require(regions.isNotEmpty()) { "Territory must have at least one region" }
    }

    val hasExclaves: Boolean get() = regions.size > 1
    val totalRingCount: Int get() = regions.sumOf { it.rings.size }
}

data class AkmRegion(val rings: List<AkmRing>) {
    init {
        require(rings.isNotEmpty()) { "Region must have at least one ring" }
    }

    val exterior: AkmRing get() = rings.first()
    val holes: List<AkmRing> get() = rings.drop(1)
    val hasHoles: Boolean get() = rings.size > 1
}

/**
 * A closed polygon ring. Points are alternating [x0, y0, x1, y1, ...] floats.
 * CCW (positive signed area) = exterior, CW (negative) = hole.
 */
data class AkmRing(val points: FloatArray) {
    init {
        require(points.size >= 6 && points.size % 2 == 0) {
            "Ring must have at least 3 points (6 floats) and even length"
        }
    }

    val pointCount: Int get() = points.size / 2
    fun x(index: Int): Float = points[index * 2]
    fun y(index: Int): Float = points[index * 2 + 1]

    fun signedArea(): Double {
        var area = 0.0
        val n = pointCount
        for (i in 0 until n) {
            val j = (i + 1) % n
            area += x(i).toDouble() * y(j).toDouble()
            area -= x(j).toDouble() * y(i).toDouble()
        }
        return area / 2.0
    }

    val isExterior: Boolean get() = signedArea() > 0
    val isHole: Boolean get() = signedArea() < 0

    override fun equals(other: Any?): Boolean =
        other is AkmRing && points.contentEquals(other.points)

    override fun hashCode(): Int = points.contentHashCode()
}

data class AkmEdge(val sourceId: UUID, val targetId: UUID, val bidirectional: Boolean) {
    init {
        require(sourceId != targetId) { "Self-referencing edges are not allowed" }
    }
}

data class AkmLayer(
    val name: String,
    val zOrder: Int,
    val blendMode: BlendMode,
    val opacity: Int,
    val imageFormat: ImageFormat,
    val imageData: ByteArray,
) {
    init {
        require(name.isNotBlank()) { "Layer name must not be blank" }
        require(opacity in 0..255) { "Opacity must be 0-255, got: $opacity" }
        require(imageData.isNotEmpty()) { "Image data must not be empty" }
    }

    val opacityFloat: Float get() = opacity / 255f

    override fun equals(other: Any?): Boolean =
        other is AkmLayer && name == other.name && zOrder == other.zOrder &&
            blendMode == other.blendMode && opacity == other.opacity &&
            imageFormat == other.imageFormat && imageData.contentEquals(other.imageData)

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + zOrder
        result = 31 * result + blendMode.hashCode()
        result = 31 * result + opacity
        result = 31 * result + imageFormat.hashCode()
        result = 31 * result + imageData.contentHashCode()
        return result
    }
}

// ============================================================================
// Enums
// ============================================================================

enum class BlendMode(val wireValue: Byte) {
    NORMAL(0), MULTIPLY(1), SCREEN(2), OVERLAY(3);

    companion object {
        fun fromWireValue(value: Byte): BlendMode =
            entries.find { it.wireValue == value }
                ?: throw IllegalArgumentException("Unknown blend mode: $value")
    }
}

enum class ImageFormat(val wireValue: Byte, val mimeType: String) {
    PNG(0, "image/png"), WEBP(1, "image/webp"), AVIF(2, "image/avif");

    companion object {
        fun fromWireValue(value: Byte): ImageFormat =
            entries.find { it.wireValue == value }
                ?: throw IllegalArgumentException("Unknown image format: $value")
    }
}

enum class AkmField(tag: String) {
    UNKNOWN("UNKN"), META("META"), GEOM("GEOM"), ADJC("ADJC"), LAYR("LAYR"), CKSM("CKSM");

    val tag: ByteArray = tag.toByteArray(StandardCharsets.UTF_8).also {
        require(it.size == 4) { "Tag must be exactly 4 bytes" }
    }

    companion object {
        fun of(tagBytes: ByteArray): AkmField =
            entries.find { it.tag.contentEquals(tagBytes) } ?: UNKNOWN
    }
}

// ============================================================================
// Encoder
// ============================================================================

object AkmEncoder {

    private val SIGNATURE = byteArrayOf(
        0x83.toByte(), 0x41, 0x4B, 0x4D,
        0x0D, 0x0A, 0x1A, 0x0A
    )

    fun encode(map: AkmMap): ByteArray {
        val out = ByteArrayOutputStream()

        // File header (16 bytes)
        out.write(SIGNATURE)
        out.writeU16(1) // version major
        out.writeU16(0) // version minor
        out.writeU32(0) // reserved

        // Chunks
        out.writeChunk(AkmField.META, encodeMeta(map))
        out.writeChunk(AkmField.GEOM, encodeGeom(map))
        out.writeChunk(AkmField.ADJC, encodeAdjc(map))
        for (layer in map.layers) {
            out.writeChunk(AkmField.LAYR, encodeLayer(layer))
        }

        // Checksum
        val contentSoFar = out.toByteArray()
        val hash = MessageDigest.getInstance("SHA-256").digest(contentSoFar)
        val cksmPayload = ByteArray(1 + hash.size)
        cksmPayload[0] = 0 // algorithm = SHA-256
        System.arraycopy(hash, 0, cksmPayload, 1, hash.size)
        out.writeChunk(AkmField.CKSM, cksmPayload)

        return out.toByteArray()
    }

    private fun encodeMeta(map: AkmMap): ByteArray = buildPayload {
        writeUuid(map.mapId)
        writeU32(map.canvasWidth)
        writeU32(map.canvasHeight)
        writeString(map.displayName)
        writeString(map.description)
        writeU16(map.authors.size)
        for (author in map.authors) {
            writeUuid(author.authorId)
            writeString(author.name)
        }
        writeI64(map.createdAt?.epochSecond ?: 0L)
        writeString(map.editorVersion)
    }

    private fun encodeGeom(map: AkmMap): ByteArray = buildPayload {
        writeU32(map.territories.size)
        for (territory in map.territories) {
            writeUuid(territory.territoryId)
            writeString(territory.label)
            writeF32(territory.centroidX)
            writeF32(territory.centroidY)
            writeU16(territory.regions.size)
            for (region in territory.regions) {
                writeU16(region.rings.size)
                for (ring in region.rings) {
                    writeU32(ring.pointCount)
                    for (v in ring.points) writeF32(v)
                }
            }
        }
    }

    private fun encodeAdjc(map: AkmMap): ByteArray = buildPayload {
        writeU32(map.edges.size)
        for (edge in map.edges) {
            writeUuid(edge.sourceId)
            writeUuid(edge.targetId)
            write(if (edge.bidirectional) 0x01 else 0x00)
        }
    }

    private fun encodeLayer(layer: AkmLayer): ByteArray = buildPayload {
        writeString(layer.name)
        writeU16(layer.zOrder)
        write(layer.blendMode.wireValue.toInt())
        write(layer.opacity)
        write(layer.imageFormat.wireValue.toInt())
        writeU32(layer.imageData.size)
        write(layer.imageData)
    }

    private inline fun buildPayload(block: ByteArrayOutputStream.() -> Unit): ByteArray =
        ByteArrayOutputStream().apply(block).toByteArray()
}

// ============================================================================
// Decoder
// ============================================================================

object AkmDecoder {

    private val SIGNATURE = byteArrayOf(
        0x83.toByte(), 0x41, 0x4B, 0x4D,
        0x0D, 0x0A, 0x1A, 0x0A
    )

    private const val HEADER_SIZE = 16

    fun decode(data: ByteArray): AkmMap {
        require(data.size >= HEADER_SIZE) { "Data too short to be a valid .akm file" }

        val input = ByteArrayInputStream(data)

        // File header
        val sig = input.readExact(8)
        if (!sig.contentEquals(SIGNATURE)) throw IOException("Invalid AKM file signature")

        val versionMajor = input.readU16()
        val versionMinor = input.readU16()
        if (versionMajor > 1) throw IOException("Unsupported AKM version: $versionMajor.$versionMinor")

        input.readExact(4) // reserved

        // Chunks
        var mapId: UUID? = null
        var canvasWidth = 0
        var canvasHeight = 0
        var displayName: String? = null
        var description = ""
        var authors: List<AkmAuthor>? = null
        var createdAt: Instant? = null
        var editorVersion = ""
        var territories: List<AkmTerritory>? = null
        var edges: List<AkmEdge>? = null
        val layers = mutableListOf<AkmLayer>()

        val checksumAccumulator = ByteArrayOutputStream()
        checksumAccumulator.write(data, 0, HEADER_SIZE)

        var foundChecksum = false

        while (input.available() > 0) {
            val tag = input.readExact(4)
            val sizeBytes = input.readExact(4)
            val payloadSize = ByteBuffer.wrap(sizeBytes).int
            if (payloadSize < 0) throw IOException("Negative chunk size")

            val payload = input.readExact(payloadSize)
            val field = AkmField.of(tag)

            if (field == AkmField.CKSM) {
                if (payload.isEmpty()) throw IOException("CKSM chunk too short")
                if (payload[0] != 0.toByte()) throw IOException("Unsupported checksum algorithm: ${payload[0]}")
                val expectedHash = payload.copyOfRange(1, payload.size)
                val actualHash = MessageDigest.getInstance("SHA-256").digest(checksumAccumulator.toByteArray())
                if (!expectedHash.contentEquals(actualHash)) throw IOException("Checksum verification failed")
                foundChecksum = true
                break
            }

            checksumAccumulator.write(tag)
            checksumAccumulator.write(sizeBytes)
            checksumAccumulator.write(payload)

            val p = ByteArrayInputStream(payload)

            when (field) {
                AkmField.META -> {
                    mapId = p.readUuid()
                    canvasWidth = p.readU32()
                    canvasHeight = p.readU32()
                    displayName = p.readString()
                    description = p.readString()
                    val authorCount = p.readU16()
                    authors = (0 until authorCount).map { AkmAuthor(p.readUuid(), p.readString()) }
                    createdAt = Instant.ofEpochSecond(p.readI64())
                    editorVersion = p.readString()
                }
                AkmField.GEOM -> {
                    val territoryCount = p.readU32()
                    territories = (0 until territoryCount).map { readTerritory(p) }
                }
                AkmField.ADJC -> {
                    val edgeCount = p.readU32()
                    edges = (0 until edgeCount).map {
                        val sourceId = p.readUuid()
                        val targetId = p.readUuid()
                        val bidirectional = (p.readExact(1)[0].toInt() and 0x01) != 0
                        AkmEdge(sourceId, targetId, bidirectional)
                    }
                }
                AkmField.LAYR -> {
                    val name = p.readString()
                    val zOrder = p.readU16()
                    val blendMode = BlendMode.fromWireValue(p.readExact(1)[0])
                    val opacity = p.readExact(1)[0].toInt() and 0xFF
                    val imageFormat = ImageFormat.fromWireValue(p.readExact(1)[0])
                    val imageDataLength = p.readU32()
                    val imageData = p.readExact(imageDataLength)
                    layers.add(AkmLayer(name, zOrder, blendMode, opacity, imageFormat, imageData))
                }
                AkmField.UNKNOWN -> { /* skip for forward compatibility */ }
                else -> { }
            }
        }

        if (!foundChecksum) throw IOException("Missing CKSM chunk")
        requireNotNull(mapId) { "Missing META chunk" }
        requireNotNull(displayName) { "Missing META chunk" }
        requireNotNull(territories) { "Missing GEOM chunk" }
        requireNotNull(edges) { "Missing ADJC chunk" }
        requireNotNull(authors) { "Missing META chunk" }

        return AkmMap(mapId, canvasWidth, canvasHeight, displayName, description,
            authors, createdAt, editorVersion, territories, edges, layers)
    }

    private fun readTerritory(input: ByteArrayInputStream): AkmTerritory {
        val territoryId = input.readUuid()
        val label = input.readString()
        val centroidX = input.readF32()
        val centroidY = input.readF32()
        val regionCount = input.readU16()
        val regions = (0 until regionCount).map {
            val ringCount = input.readU16()
            val rings = (0 until ringCount).map {
                val pointCount = input.readU32()
                val points = FloatArray(pointCount * 2) { input.readF32() }
                AkmRing(points)
            }
            AkmRegion(rings)
        }
        return AkmTerritory(territoryId, label, centroidX, centroidY, regions)
    }
}

// ============================================================================
// Extension functions for binary I/O
// ============================================================================

private fun ByteArrayInputStream.readExact(n: Int): ByteArray {
    val buf = readNBytes(n)
    if (buf.size != n) throw IOException("Unexpected end of file (expected $n bytes, got ${buf.size})")
    return buf
}

private fun ByteArrayInputStream.readU16(): Int {
    val b = readExact(2)
    return ((b[0].toInt() and 0xFF) shl 8) or (b[1].toInt() and 0xFF)
}

private fun ByteArrayInputStream.readU32(): Int = ByteBuffer.wrap(readExact(4)).int
private fun ByteArrayInputStream.readI64(): Long = ByteBuffer.wrap(readExact(8)).long
private fun ByteArrayInputStream.readF32(): Float = ByteBuffer.wrap(readExact(4)).float

private fun ByteArrayInputStream.readUuid(): UUID {
    val buf = ByteBuffer.wrap(readExact(16))
    return UUID(buf.long, buf.long)
}

private fun ByteArrayInputStream.readString(): String {
    val len = readU16()
    if (len == 0) return ""
    return String(readExact(len), StandardCharsets.UTF_8)
}

private fun ByteArrayOutputStream.writeU16(value: Int) {
    write((value shr 8) and 0xFF)
    write(value and 0xFF)
}

private fun ByteArrayOutputStream.writeU32(value: Int) {
    write(ByteBuffer.allocate(4).putInt(value).array())
}

private fun ByteArrayOutputStream.writeI64(value: Long) {
    write(ByteBuffer.allocate(8).putLong(value).array())
}

private fun ByteArrayOutputStream.writeF32(value: Float) {
    write(ByteBuffer.allocate(4).putFloat(value).array())
}

private fun ByteArrayOutputStream.writeUuid(uuid: UUID) {
    val buf = ByteBuffer.allocate(16)
    buf.putLong(uuid.mostSignificantBits)
    buf.putLong(uuid.leastSignificantBits)
    write(buf.array())
}

private fun ByteArrayOutputStream.writeString(value: String) {
    if (value.isEmpty()) {
        writeU16(0)
        return
    }
    val bytes = value.toByteArray(StandardCharsets.UTF_8)
    writeU16(bytes.size)
    write(bytes)
}

private fun ByteArrayOutputStream.writeChunk(field: AkmField, payload: ByteArray) {
    write(field.tag)
    writeU32(payload.size)
    write(payload)
}
